File: IndexAccess/AdvancedIndexReader.cpp
#include "AdvancedIndexReader.h"
#include "IndexContext.h"
#include <algorithm>
#include <cassert>
#include <cstring>
#include <cmath>

namespace {

float StreamWeight(char stream, const QueryCompileModeParameters& parameters)
{
    switch (stream) {
    case 'A': return parameters.QMP_AnchorWeight;
    case 'U': return parameters.QMP_UrlWeight;
    case 'T': return parameters.QMP_TitleWeight;
    case 'B': return parameters.QMP_BodyWeight;
    default: return 1.0f;
    }
}

}

void AdvancedIndexReader::Open(const char*      streamKey,
                                IndexBlockTable* blockTable,
                                const IndexContext* context,
                                uint32_t wordSpan)
{
    assert(streamKey);
    assert(blockTable);
    assert(context);

    if (m_BlockTable && m_BlockSlotNumber != UINT32_MAX)
        m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);

    delete[] m_Word;
    m_Word = new char[std::strlen(streamKey) + 1];
    std::strcpy(m_Word, streamKey);
    m_Stream = m_Word[std::strlen(m_Word) - 1];
    m_SourceMask = ReaderSourceMaskForStream(m_Stream);

    m_BlockTable      = blockTable;
    m_Context         = context;
    m_DocFreq         = 0;
    m_WordSpan        = std::max(1u, wordSpan);
    m_SpanWeight      = m_Context->GetSpanWeight(m_WordSpan) * StreamWeight(m_Stream, m_Context->GetQueryParameters());
    m_Idf             = 0.0f;
    m_Bm25LengthBias  = 0.0f;
    m_Bm25LengthScale = 0.0f;
    m_BlockSeqNumber  = 0;
    m_BlockSlotNumber = UINT32_MAX;
    m_RemainingContinuationBlocks = 0;

    uint32_t indexOffset = 0, indexLength = 0;

    bool found = m_BlockTable->FindTermData(streamKey,
                                            &m_BlockSeqNumber, &indexOffset,
                                            &indexLength,      &m_DocFreq,
                                            &m_RemainingContinuationBlocks);
    if (found) {
        static constexpr float K1 = 1.2f;
        static constexpr float B = 0.75f;

        const IndexFileHeader& header = m_Context->GetIndexFileHeader();
        assert(header.IFH_NumDocuments > 0);
        assert(header.IFH_AvgDocLength > 0.0f);

        const float totalDocs = static_cast<float>(header.IFH_NumDocuments);
        const float df = static_cast<float>(std::max(1u, m_DocFreq));
        m_Idf = std::max(0.0f,
            std::log((totalDocs - df + 0.5f) / (df + 0.5f) + 1.0f));
        m_Bm25LengthBias = K1 * (1.0f - B);
        m_Bm25LengthScale = K1 * B / std::max(1.0f, m_Context->GetAverageStreamLength(m_Stream));

        IndexBlock* block = reinterpret_cast<IndexBlock*>(
            m_BlockTable->GetBlock(BlockKind::Index, m_BlockSeqNumber, &m_BlockSlotNumber));
        assert(block);
        m_Decoder.OpenRaw(block->IB_Data + indexOffset, indexLength);
    }

    GoNext();
}

void AdvancedIndexReader::GoNext()
{
    m_Decoder.GoNext();

    while (m_Decoder.IsEnd() && m_RemainingContinuationBlocks > 0) {
        m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);
        IndexBlock* next = reinterpret_cast<IndexBlock*>(
            m_BlockTable->GetBlock(BlockKind::Index, m_BlockSeqNumber + 1, &m_BlockSlotNumber));
        assert(next);
        ++m_BlockSeqNumber;
        const auto* header = reinterpret_cast<const IndexBlockContinuationHeader*>(next->IB_Data);
        m_Decoder.OpenRaw(next->IB_Data + sizeof(IndexBlockContinuationHeader), header->IBCH_DataLength);
        --m_RemainingContinuationBlocks;
        m_Decoder.GoNext();
    }

    if (m_Decoder.IsEnd() && m_RemainingContinuationBlocks == 0) {
        m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);
        m_BlockSlotNumber = UINT32_MAX;
    }
}

void AdvancedIndexReader::GoUntil(uint64_t target, uint64_t /*limit*/)
{
    while (true) {
        m_Decoder.GoUntil(target);
        if (!m_Decoder.IsEnd()) break;
        if (m_RemainingContinuationBlocks == 0) {
            m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);
            m_BlockSlotNumber = UINT32_MAX;
            break;
        }
        m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);
        IndexBlock* next = reinterpret_cast<IndexBlock*>(
            m_BlockTable->GetBlock(BlockKind::Index, m_BlockSeqNumber + 1, &m_BlockSlotNumber));
        assert(next);
        const auto* header = reinterpret_cast<const IndexBlockContinuationHeader*>(next->IB_Data);
        ++m_BlockSeqNumber;
        --m_RemainingContinuationBlocks;
        if (target > header->IBCH_MaxDocID) {
            m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);
            m_BlockSlotNumber = UINT32_MAX;
            continue;
        }
        m_Decoder.OpenRaw(next->IB_Data + sizeof(IndexBlockContinuationHeader), header->IBCH_DataLength);
        m_Decoder.GoNext();
        if (m_Decoder.IsEnd()) {
            m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);
            m_BlockSlotNumber = UINT32_MAX;
            break;
        }
    }
}

bool AdvancedIndexReader::IsEnd()          { return m_Decoder.IsEnd(); }
uint64_t AdvancedIndexReader::GetDocumentID() {
    return m_Decoder.GetDocumentID();
}
uint32_t AdvancedIndexReader::GetTermFreq() {
    return m_Decoder.GetTermFreq();
}
float AdvancedIndexReader::GetScore(const DocDataEntry* entry) {
    assert(entry);
    const uint32_t docLength = IndexContext::GetStreamLength(*entry, m_Stream);
    assert(docLength > 0);

    const float tf = static_cast<float>(m_Decoder.GetTermFreq());
    const float dl = static_cast<float>(std::max(1u, docLength));

    static constexpr float K1PlusOne = 2.2f;
    return m_SpanWeight * m_Idf * tf * K1PlusOne /
        (tf + m_Bm25LengthBias + m_Bm25LengthScale * dl);
}
void AdvancedIndexReader::Close() {
    if (m_BlockTable && m_BlockSlotNumber != UINT32_MAX)
        m_BlockTable->ReleaseBlock(BlockKind::Index, m_BlockSlotNumber);
    m_BlockSlotNumber = UINT32_MAX;
    delete[] m_Word;
    m_Word = nullptr;
}

File: RustBlade/src/IndexAccess/AdvancedIndexReader.rs
use std::sync::Arc;

use crate::block_table::{
    BlockHandle,
    BlockKind,
    IndexBlock,
    IndexBlockContinuationHeader,
    IndexBlockTable,
    INDEX_BLOCK_CONTINUATION_HEADER_SIZE,
};
use crate::block_table::DocDataEntry;
use crate::index_reader::{IndexReader, ReaderSourceMaskForStream};
use crate::unified_decoder::UnifiedDecoder;

#[allow(non_snake_case)]
pub struct AdvancedIndexReader {
    #[allow(dead_code)]
    m_Word: String,
    m_BlockTable: Arc<IndexBlockTable>,
    m_DocFreq: u32,
    m_Stream: char,
    m_SourceMask: u8,
    #[allow(dead_code)]
    m_WordSpan: u32,
    m_SpanWeight: f32,
    m_Idf: f32,
    m_Bm25LengthBias: f32,
    m_Bm25LengthScale: f32,
    m_BlockSeqNumber: u32,
    m_RemainingContinuationBlocks: u32,
    m_Decoder: UnifiedDecoder,
    m_Debug: bool,
    m_DebugDepth: usize,
}

#[allow(non_snake_case)]
impl AdvancedIndexReader {
    pub fn Open(
        stream_key: &str,
        block_table: Arc<IndexBlockTable>,
        num_documents: u64,
        average_stream_length: f32,
        span_weight: f32,
        word_span: u32,
    ) -> Self {
        let mut reader = Self {
            m_Word: stream_key.to_string(),
            m_BlockTable: block_table,
            m_DocFreq: 0,
            m_Stream: stream_key.chars().last().unwrap_or('B'),
            m_SourceMask: stream_key.chars().last().map(ReaderSourceMaskForStream).unwrap_or(0),
            m_WordSpan: word_span.max(1),
            m_SpanWeight: span_weight,
            m_Idf: 0.0,
            m_Bm25LengthBias: 0.0,
            m_Bm25LengthScale: 0.0,
            m_BlockSeqNumber: 0,
            m_RemainingContinuationBlocks: 0,
            m_Decoder: UnifiedDecoder::new(),
            m_Debug: false,
            m_DebugDepth: 0,
        };

        if let Some(location) = reader.m_BlockTable.FindTermData(stream_key) {
            reader.m_BlockSeqNumber = location.index_block_id;
            reader.m_RemainingContinuationBlocks = location.continuation_block_count;
            reader.m_DocFreq = location.doc_freq;
            const K1: f32 = 1.2;
            const B: f32 = 0.75;
            assert!(num_documents > 0);
            assert!(average_stream_length > 0.0);
            let totalDocs = num_documents as f32;
            let docFreq = reader.m_DocFreq.max(1) as f32;
            reader.m_Idf = (((totalDocs - docFreq + 0.5) / (docFreq + 0.5)) + 1.0).ln().max(0.0);
            reader.m_Bm25LengthBias = K1 * (1.0 - B);
            reader.m_Bm25LengthScale = K1 * B / average_stream_length.max(1.0);
            if let Some(block) = reader.m_BlockTable.GetBlock(BlockKind::Index, location.index_block_id, false) {
                reader.m_Decoder.OpenRaw(block, location.index_offset, location.index_length);
            }
        }

        reader.GoNext();
        reader
    }

    fn LoadContinuation(&mut self) -> bool {
        let nextSeq = self.m_BlockSeqNumber + 1;
        self.ReleaseCurrentBlock();
        let Some(block) = self.m_BlockTable.GetBlock(BlockKind::Index, nextSeq, false) else { return false; };
        self.m_BlockSeqNumber = nextSeq;
        self.OpenContinuation(block);
        self.m_RemainingContinuationBlocks = self.m_RemainingContinuationBlocks.saturating_sub(1);
        true
    }

    fn OpenContinuation(&mut self, block: BlockHandle<IndexBlock>) {
        if let Some(header) = IndexBlockContinuationHeader::from_bytes(&block.IB_Data) {
            let len = header.IBCH_DataLength as usize;
            self.m_Decoder.OpenRaw(
                block,
                INDEX_BLOCK_CONTINUATION_HEADER_SIZE,
                len);
        } else {
            self.m_Decoder.OpenRaw(block, 0, 0);
        }
    }

    fn ReleaseCurrentBlock(&mut self) {
        if let Some(block) = self.m_Decoder.TakeBlock() {
            self.m_BlockTable.ReleaseBlock(block.Kind(), block.Slot(), false);
        }
    }
}

impl Drop for AdvancedIndexReader {
    fn drop(&mut self) { self.ReleaseCurrentBlock(); }
}

impl IndexReader for AdvancedIndexReader {
    fn GoNext(&mut self) {
        self.m_Decoder.GoNext();
        while self.m_Decoder.IsEnd() && self.m_RemainingContinuationBlocks > 0 {
            if !self.LoadContinuation() { break; }
            self.m_Decoder.GoNext();
        }
        if self.m_Decoder.IsEnd() && self.m_RemainingContinuationBlocks == 0 {
            self.ReleaseCurrentBlock();
            self.m_Decoder.Close();
        }
    }

    fn GoUntil(&mut self, target: u64, _limit: u64) {
        loop {
            self.m_Decoder.GoUntil(target);
            if !self.m_Decoder.IsEnd() { break; }
            if self.m_RemainingContinuationBlocks == 0 {
                self.ReleaseCurrentBlock();
                break;
            }
            self.ReleaseCurrentBlock();
            let Some(block) = self.m_BlockTable.GetBlock(BlockKind::Index, self.m_BlockSeqNumber + 1, false) else { break; };
            let Some(header) = IndexBlockContinuationHeader::from_bytes(&block.IB_Data) else {
                self.m_BlockTable.ReleaseBlock(block.Kind(), block.Slot(), false);
                break;
            };
            self.m_BlockSeqNumber += 1;
            self.m_RemainingContinuationBlocks = self.m_RemainingContinuationBlocks.saturating_sub(1);
            if target > header.IBCH_MaxDocID {
                self.m_BlockTable.ReleaseBlock(block.Kind(), block.Slot(), false);
                continue;
            }
            self.OpenContinuation(block);
            self.m_Decoder.GoNext();
            if self.m_Decoder.IsEnd() {
                self.ReleaseCurrentBlock();
                break;
            }
        }
        if self.m_Decoder.IsEnd() && self.m_RemainingContinuationBlocks == 0 {
            self.ReleaseCurrentBlock();
            self.m_Decoder.Close();
        }
    }

    fn IsEnd(&self) -> bool { self.m_Decoder.IsEnd() }

    fn GetDocumentID(&self) -> u64 {
        self.m_Decoder.GetDocumentID()
    }

    fn GetTermFreq(&self) -> u32 {
        self.m_Decoder.GetTermFreq() as u32
    }

    fn GetScore(&mut self, entry: &DocDataEntry) -> f32 {
        const K1_PLUS_ONE: f32 = 2.2;
        let tf = self.GetTermFreq() as f32;
        let docLength = match self.m_Stream {
            'T' => entry.DDE_TitleLength,
            'B' => entry.DDE_BodyLength,
            'U' => entry.DDE_UrlLength,
            'A' => entry.DDE_AnchorLength,
            'M' => entry.DDE_MetaLength,
            _ => 0,
        };
        let docLength = if docLength > 0 { docLength } else { entry.DDE_BodyLength.max(1) } as f32;
        self.m_SpanWeight * self.m_Idf * tf * K1_PLUS_ONE /
            (tf + self.m_Bm25LengthBias + self.m_Bm25LengthScale * docLength)
    }

    fn GetSourceMask(&mut self) -> u8 { self.m_SourceMask }

    fn SetDebug(&mut self, _label: &str, depth: usize) {
        self.m_Debug = true;
        self.m_DebugDepth = depth;
        if !self.m_Decoder.IsEnd() {
            println!("{}[leaf] {:<12}  -{}-", " ".repeat(depth * 2), self.m_Word, self.m_Decoder.GetDocumentID());
        } else {
            println!("{}[leaf] {:<12}  (empty)", " ".repeat(depth * 2), self.m_Word);
        }
    }
    fn Close(&mut self) {
        self.ReleaseCurrentBlock();
        self.m_Decoder.Close();
        self.m_Word.clear();
    }
}
